server/venueless/core/models/auth.py

```diff
@@ -19,3 +19,34 @@ def serialize_public(self):
         # Important: If this is updated, venueless.core.services.user.get_public_users also needs to be updated!
         # For performance reasons, it does not use this method directly.
         return {"id": str(self.id), "profile": self.profile}
+
+    def get_role_grants(self, room=None):
+        roles = set(self.world_grants.values_list("role", flat=True))
+        if room:
+            roles |= set(
+                self.room_grants.filter(room=room).values_list("role", flat=True)
+            )
+        return roles
+
+
+class RoomGrant(models.Model):
+    world = models.ForeignKey(
+        "World", related_name="room_grants", on_delete=models.CASCADE
+    )
+    room = models.ForeignKey(
+        "Room", related_name="role_grants", on_delete=models.CASCADE
+    )
+    user = models.ForeignKey(
+        "User", related_name="room_grants", on_delete=models.CASCADE
+    )
+    role = models.CharField(max_length=200)
+
+
+class WorldGrant(models.Model):
```

Comment:
Hmm, I think it would be too much. First, generic foreign keys are nasty. Second, we wouldn't just have to annotate proper ``@require_world_permission`` and ``@room_action`` decorators, we'd instead need to think about permission checks for every single thing. Note that I also dropped permission from the module config level (which was previously in the design) to have them just on world/room level to keep it simpler.